interpreter/host/js.ml

```diff
@@ -198,14 +218,30 @@ let assert_return lits ts at =
   in [], List.flatten (List.rev_map test lits)
 
 let assert_return_nan ts at =
-  let var i = Int32.of_int i @@ at in
-  let init i t = [SetLocal (var i) @@ at] in
-  let test i t =
-    [ GetLocal (var i) @@ at;
-      GetLocal (var i) @@ at;
-      Compare (eq_of t) @@ at;
+  let test t =
+    let t', reinterpret = reinterpret_of t in
+    [ reinterpret @@ at;
+      (* it can be any NaN that's one everywhere the canonical NaN is one *)
+      Const (canonical_nan_of t' @@ at) @@ at;
+      Binary (and_of t') @@ at;
+      Const (canonical_nan_of t' @@ at) @@ at;
+      Compare (eq_of t') @@ at;
+      Test (Values.I32 I32Op.Eqz) @@ at;
+      BrIf (0l @@ at) @@ at ]
+  in [], List.flatten (List.map test ts)
+
+let assert_return_canonical_nan ts at =
```

Comment:
The only difference between the two functions is the initial mask AFAICS. How about making that a parameter?